src/arrow1/reader_async.rs

```diff
@@ -0,0 +1,33 @@
+use crate::arrow1::error::Result;
+use crate::common::fetch::{create_reader, get_content_length};
+use crate::log;
+
+use arrow::ipc::writer::StreamWriter;
+use arrow::record_batch::RecordBatchReader;
+use futures::stream::StreamExt;
+use parquet::arrow::async_reader::ParquetRecordBatchStreamBuilder;
+
+use async_compat::CompatExt;
+
+pub async fn read_parquet(url: String) -> Result<Vec<u8>> {
+    let length = get_content_length(url.clone()).await?;
+    let reader = create_reader(url, length, None);
+    let builder = ParquetRecordBatchStreamBuilder::new(reader.compat()).await?;
+    // quite a few options here - projection masks, row group subselection, etc...
+    let arrow_schema = builder.schema().clone();
+    let parquet_reader = builder.build()?;
+    let intermediate: Vec<_> = parquet_reader.collect().await;
```

Comment:
Yep, this one does pretty much just do a regular readParquet - we would need to use a similar wrapper struct as in arrow2.

I think for both cases we could mirror some of the internal Builder/Reader structures - with a couple of main usage scenarios:
1. First-K row groups or all row groups - reader.stream(). Consumption can be paused, prematurely stopped without too much hassle (it's all platform-level). The Geoarrow + parquet-wasm observable is what got me thinking about this in the first place (you still want the whole file, but the transform you're applying to it is taxing enough to warrant a stream, which gets you built-in backpressure).
2. Skip a few row groups, pick specific indexes - closer to the current functional approach, you loop over the desired row group indices and dispatch read_row_group calls as usual. Arrow1 may have a significant advantage here, since it looks like it has well-defined filtering specs, at several levels of granularity (down to the PageIndex apparently).